### DISCUSSION TITLE
Clear/Reset State After User Sign Out

### DISCUSSION THREAD

**Discussion by willcray** [pmndrs/zustand#574] [Q&A] [answered]
What's the proper way to reset a store? 

When a user logs out in my app, I want to reset all the store state variables, because they contain information specific to the previous user. 

I want to reset the state variables to their initial values.

**Comment by dai-shi** [comment, marked as answer]
keep the initial state (excluding functions) and overwrite?
```js
const initialState = {
  count: 10,
}
const useStore = create((set) => ({
  ...initialState,
  inc: () => set(state => ({ count: state.count + 1 })),
  reset: () => set(initialState),
}))
```

**Reply by HelloRickey** [reply]
awesome!!!
What a genius design!

**Comment by ntlongctt** [comment]
@dai-shi thank you for your suggestion, but how about if I have multiple stores in some places, then I want to reset them all once? I'm trying to write a wrapper function

```jsx
export const useStoreWithPersis = <T extends State>({
  initialState,
  key,
}: StoreWithPersist<T>) => {
  const _store = create<T>()(
    persist(
      () => ({
        ...initialState,
      }),
      { name: key },
    ),
  );

  if (!persistReseter[key]) {
    persistReseter[key] = () => _store.setState(initialState, true);
  }

  return _store;
};
```
then write a resetAll function like that:
```jsx
export const resetAllSlices = () => {
  for (const resetKey in persistReseter) {
    localStorage.removeItem(resetKey);
    persistReseter[resetKey]();
  }
};
```

it's working fine, but as you can see I need to create a new wrapper when need use another middleware. Is there any way to solve this problem

**Reply by dai-shi** [reply]
It looks good to me (except for the function name).

You can create a higher order function:
```js
const resetSet = new Set();
const resettable = (create) => (initialState) => {
  const useStore = create(() => initialState);
  const reset = () => useStore.setState(initialState, true);
  resetSet.add(reset);
  return useStore;
};
const resetAll = () => resetSet.forEach((reset) => reset());
```
But, it's probably hard to type it. I think your approach is easier.

**Reply by zhongjiaguo001** [reply]
How to reset all stores at once with persistence plugin in Zustand for optimized text

**Reply by dai-shi** [reply]
I'm not sure, but the rehydrate api might help?
https://github.com/pmndrs/zustand/blob/0058a03b78bf628a27fb3cb4b6ebe3e6263dd0a0/docs/integrations/persisting-store-data.md?plain=1#L361-L370

**Reply by zhongjiaguo001** [reply]
Thank you
